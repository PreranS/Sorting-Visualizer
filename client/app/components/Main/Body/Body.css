#bodyContainer {
  width: 100%;
  height: 100%;
  background-color: rgb(230, 230, 230);
  text-align: center;
  display: flex;
  flex-direction: row;
  align-items: flex-end;
  justify-content: center;
  padding: 24px 0 0 0;
  flex-wrap: nowrap;
  min-height: 0;
  height: 100%;
  overflow-x: hidden;
}

.arrayElement {
  padding-top: 7px;
  font-family: sans-serif;
  font-weight: 700;
  display: flex;
  align-items: flex-end;
  justify-content: center;
  border-radius: 6px 6px 0 0;
  margin-bottom: 0;
  box-shadow: 0 2px 8px rgba(0,0,0,0.08);
  transition: background 0.3s, height 0.3s, color 0.2s;
  user-select: none;
  flex: 1 1 0;
  min-width: 2px;
  overflow: hidden;
}

/* Bar color states */
.arrayElement.swapper {
  background-color: rgba(219, 57, 57, 0.8);
}
.arrayElement.active {
  background-color: rgba(78, 216, 96, 0.8);
}
.arrayElement.pivot {
  background-color: rgba(237, 234, 59, 0.8);
}
.arrayElement.sorted {
  background-color: rgba(169, 92, 232, 0.8);
}
.arrayElement.defaultBar {
  background-color: rgba(66, 134, 244, 0.8);
}

/* Hide numbers for very thin bars */
.arrayElement {
  color: inherit;
}
.arrayElement[style*='width: 2px'],
.arrayElement[style*='width: 3px'],
.arrayElement[style*='width: 4px'] {
  color: transparent !important;
}

/* Responsive design for mobile */
@media (max-width: 600px) {
  #bodyContainer {
    min-height: 180px;
    padding: 12px 0 0 0;
  }
  .arrayElement {
    font-size: 10px !important;
    padding-top: 3px;
    border-radius: 4px 4px 0 0;
  }
}
