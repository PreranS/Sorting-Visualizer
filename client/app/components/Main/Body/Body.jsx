import React, { Component } from "react";
import "./Body.css";

class Body extends Component {
  constructor(props) {
    super(props);
  }

  render() {
    const {
      array,
      currentBubbleTwo,
      currentQuickTwo,
      pivot,
      currentSwappers,
      currentHeapThree,
      currentSorted,
      currentMergeX,
    } = this.props;

    const numWidth = Math.floor(window.innerWidth / (array.length * 3));
    const width = `${numWidth}px`;
    const numMargin = array.length < 5 ?
      10 : array.length < 8 ?
        8 : array.length < 11 ?
          6 : array.length < 20 ?
            4 : array.length < 50 ?
              3.5 : array.length < 100 ?
                3 : array.length < 130 ?
                  2.5 : 2;
    const margin = `${numMargin}px`;
    const color = numWidth > 20 ? "white" : "transparent";
    const numFont = numWidth > 70 ?
      20 : numWidth > 60 ?
        18 : numWidth > 50 ?
          16 : numWidth > 40 ?
            14 : numWidth > 30 ?
              12 : numWidth > 20 ?
                10 : 8;
    const fontSize = `${numFont}px`

    return (
      <div id="bodyContainer" role="region" aria-label="Sorting Visualization">
        { array.length ? array.map((number, index) => {
          let barClass = "arrayElement ";
          if (currentSwappers.includes(index)) barClass += "swapper ";
          else if (currentBubbleTwo.includes(index) || currentQuickTwo.includes(index) || currentHeapThree.includes(index) || currentMergeX.includes(index)) barClass += "active ";
          else if (pivot === index) barClass += "pivot ";
          else if (currentSorted.includes(index)) barClass += "sorted ";
          else barClass += "defaultBar ";
          return <div
            className={barClass}
            key={index}
            style={{height: `${number * 3}px`, width: width, marginLeft: margin, marginRight: margin, color: color, fontSize: fontSize}}
            aria-label={`Array value ${number}`}
            title={`Value: ${number}`}
          >
            {number}
          </div>
        }) : null
        }
      </div>
    )
  }
}

export default Body;
