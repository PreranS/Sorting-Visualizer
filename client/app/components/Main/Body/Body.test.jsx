import React from 'react';
import renderer from 'react-test-renderer';
import Body from './Body';

test('renders Body', () => {
  const mockArray = [1, 2, 3, 4, 5];
  const mockProps = {
    array: mockArray,
    currentSwappers: [],
    currentBubbleTwo: [],
    currentQuickTwo: [],
    currentHeapThree: [],
    currentMergeX: [],
    pivot: null,
    currentSorted: [],
  };
  const tree = renderer.create(<Body {...mockProps} />).toJSON();
  expect(tree).toMatchSnapshot();
});