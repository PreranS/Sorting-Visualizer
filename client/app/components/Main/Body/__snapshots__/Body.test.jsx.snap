// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`renders Body 1`] = `
<div
  aria-label="Sorting Visualization"
  id="bodyContainer"
  role="region"
>
  <div
    aria-label="Array value 1"
    className="arrayElement defaultBar "
    style={
      {
        "color": "white",
        "fontSize": "18px",
        "height": "3px",
        "marginLeft": "8px",
        "marginRight": "8px",
        "width": "68px",
      }
    }
    title="Value: 1"
  >
    1
  </div>
  <div
    aria-label="Array value 2"
    className="arrayElement defaultBar "
    style={
      {
        "color": "white",
        "fontSize": "18px",
        "height": "6px",
        "marginLeft": "8px",
        "marginRight": "8px",
        "width": "68px",
      }
    }
    title="Value: 2"
  >
    2
  </div>
  <div
    aria-label="Array value 3"
    className="arrayElement defaultBar "
    style={
      {
        "color": "white",
        "fontSize": "18px",
        "height": "9px",
        "marginLeft": "8px",
        "marginRight": "8px",
        "width": "68px",
      }
    }
    title="Value: 3"
  >
    3
  </div>
  <div
    aria-label="Array value 4"
    className="arrayElement defaultBar "
    style={
      {
        "color": "white",
        "fontSize": "18px",
        "height": "12px",
        "marginLeft": "8px",
        "marginRight": "8px",
        "width": "68px",
      }
    }
    title="Value: 4"
  >
    4
  </div>
  <div
    aria-label="Array value 5"
    className="arrayElement defaultBar "
    style={
      {
        "color": "white",
        "fontSize": "18px",
        "height": "15px",
        "marginLeft": "8px",
        "marginRight": "8px",
        "width": "68px",
      }
    }
    title="Value: 5"
  >
    5
  </div>
</div>
`;
